File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Messages/SendMessageUseCase.swift
import Combine
import DI
import Foundation
import Models
import Repositories

@Factory
public final class SendMessageUseCase {
    private let messagesRepository: MessagesRepository

    public func callAsFunction(dog: Dog, text: String) -> AnyPublisher<Void, MessageError> {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return switch trimmed.isEmpty {
        case true: Fail(error: MessageError.emptyMessage).eraseToAnyPublisher()
        case false: messagesRepository.addMessage(dog: dog, text: trimmed)
        }
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/messages/SendMessageUseCase.kt
package com.perrystreet.woof.usecase.messages

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.models.errors.MessageException
import com.perrystreet.woof.repositories.messages.MessagesRepository
import io.reactivex.rxjava3.core.Completable
import org.koin.core.annotation.Factory

@Factory
class SendMessageUseCase(
    private val messagesRepository: MessagesRepository,
) {
    operator fun invoke(dog: Dog, text: String): Completable {
        val trimmed = text.trim()
        return when (trimmed.isEmpty()) {
            true -> Completable.error(MessageException.EmptyMessage)
            false -> messagesRepository.addMessage(dog, trimmed)
        }
    }
}
